src/server/pfs/server/driver.go

```diff
@@ -2119,6 +2121,44 @@ func (d *driver) addBranchProvenance(txnCtx *txncontext.TransactionContext, bran
 	return errors.EnsureStack(err)
 }
 
+func (d *driver) deleteProjectsRepos(ctx context.Context, projects []*pfs.Project) ([]*pfs.Repo, error) {
+	var repos, deleted []*pfs.Repo
+
+	var projectFilter = make(map[string]bool)
+	for _, project := range projects {
+		if project.Name == "" {
+			return nil, status.Error(codes.InvalidArgument, "empty project name")
+		}
+		projectFilter[project.Name] = true
+	}
+	if err := d.listRepo(ctx, false, "", projectFilter, func(repoInfo *pfs.RepoInfo) error {
+		repos = append(repos, repoInfo.Repo)
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if len(repos) == 0 {
+		return nil, nil
+	}
+
+	if err := d.txnEnv.WithWriteContext(ctx, func(txnCtx *txncontext.TransactionContext) error {
+		// the list does not use the transaction
```

Comment:
Ok I'm missing something. Why not do the whole operation in a transaction?

Reply:
It’s the same pattern as `deleteAll` below.

Reply:
I think this should be in a single transaction to support all-or-none semantics

Reply:
so I actually re-implemented this function in the provenance PR:
https://github.com/pachyderm/pachyderm/pull/8306/files#diff-a40f20848be655d50f1da1f38f4232133ba02eb4e63e01099055003740955d14R343
In short, we don't need the pachyderm API to support transactions to make operations transactional as long as our internal APIs support transactions.

I'm going to have to merge these changes in anyway, so it's fine if we go with this for now